src/packages/client/src/runtime/getPrismaClient.ts

```diff
@@ -907,15 +935,107 @@ new PrismaClient({
       )
     }
 
-    async $transaction(promises: Array<any>): Promise<any> {
+    /**
+     * @deprecated
+     */
+    async $___transaction(promises: Array<any>): Promise<any> {
       try {
-        return this.$transactionInternal(promises)
+        return this.$___transactionInternal(promises)
       } catch (e) {
         e.clientVersion = this._clientVersion
         throw e
       }
     }
 
+    /**
+     * Execute queries within a transaction
+     * @param input a callback or a query list
+     * @param options to set timeouts
+     * @returns
+     */
+    $transaction(input: any, options?: any) {
+      // eslint-disable-next-line prettier/prettier
+      if (!this._hasPreviewFlag('longRunningTransactions')) {
+        return this.$___transaction(input)
+      }
+
+      try {
+        return this._transaction(input, options)
+      } catch (e) {
+        e.clientVersion = this._clientVersion
+
+        throw e
+      }
+    }
+
+    /**
+     * Decide upon which transaction logic to use
+     * @param input
+     * @param options
+     * @returns
+     */
+    private async _transaction(input: any, options?: any) {
+      if (typeof input === 'function') {
+        return this._transactionWithCallback(input, options)
+      }
+
+      return this._transactionWithRequests(input, options)
+    }
+
+    /**
+     * Perform a long-running transaction
+     * @param callback
+     * @param options
+     * @returns
+     */
+    private async _transactionWithCallback(
+      callback: (client: Client) => Promise<unknown>,
+      options?: { maxWait: number; timeout: number },
+    ) {
+      // we ask the query engine to open a transaction
+      const info = await this._engine.transaction('start', options)
```

Comment:
I imagine that this takes a connection for the DB pool when it opens the connection. What happens if that connection is bad? Currently we can retry by having a retry middleware but this middleware wont be called for this start thing, so I imagine this will just throw.
For us this means I need to change all places in code that uses transaction to add retry logic.
I think the retry of failed connection needs some thought put into it.

Reply:
That's what `await this.start` just above does. Is that what you meant?

Reply:
Marking this resolved in favor of the top conversation https://github.com/prisma/prisma/pull/8384#discussion_r680353340

Reply:
Agreed